File: [PGS]/level2/네트워크.java
import java.util.*;

class Solution {
    public static ArrayList<ArrayList<Integer>> list = new ArrayList<ArrayList<Integer>>();
    public static boolean[] visited;
    public int solution(int n, int[][] computers) {
        int answer = 0;
        for(int i=0; i<=n ;i++){
            list.add(new ArrayList<Integer>());
        }
        
        for(int i=0; i<computers.length; i++){
            for(int j=0; j<n; j++){
                if(computers[i][j] == 1 && i != j){
                    list.get(i+1).add(j+1);
                }
            }
        }
        visited = new boolean[n+1];
        for(int i=1; i<=n; i++){
            if(DFS(i) == true){
                answer += 1;
            }
        }
        return answer;
    }
    public static boolean DFS(int v){
        if(visited[v] == true){
            return false;
        }
        else{
            visited[v] = true;
            for(int i=0; i<list.get(v).size(); i++){
                DFS(list.get(v).get(i));
            }
            return true;
        }
    }
}

File: [PGS]/level2/네트워크.py
def solution(n, computers):
    answer = 0
    
    parent = [0] * (n)
    
    for i in range(n):
        parent[i] = i
    
    for i in range(n):
        for j in range(n):
            if i == j :
                continue
            if computers[i][j] == 1:
                union_parent(parent, i, j)
    li = []
    for i in range(n):
        v = find_parent(parent, i)
        if not v in li:
            li.append(v)
    
    return len(li)

def find_parent(parent, x):
    if parent[x] != x:
        parent[x] = find_parent(parent, parent[x])
    return parent[x]

def union_parent(parent, a, b):
    a = find_parent(parent, a)
    b = find_parent(parent, b)
    if a < b:
        parent[b] = a
    else:
        parent[a] = b
